### ISSUE TITLE
Get IoU not mAP50 or mAP50-95

### ISSUE THREAD

**Issue by ovidedecroly** [ultralytics/ultralytics#22549]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi @glenn-jocher  I was asked to display IoU not from mAP50 and mAP50-95, how to do it for YOLOv8 and RT-DETR? Thanks

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ovidedecroly, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get started quickly — an Ultralytics engineer will also assist you here soon. We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by ovidedecroly** [comment]
Hi @glenn-jocher 
I was asked to display IoU not from mAP50 and mAP50-95, how to do it for YOLOv8 and RT-DETR? Thanks

# ==============================================================
# 🧩 CODE V6 (Stabil) — IoU Auto-Match (Logic Fix)
# ==============================================================
# ✅ Tujuan:
# 1. Menggunakan 'ValidatorClass' (API level rendah) yang kompatibel.
# 2. Mengakses 'validator.data['test']' (struktur fork planktovide).
# 3. Menurunkan path 'test/labels' secara manual.
# 4. Menghitung IoU dan menyimpan hasil.
# ==============================================================

import os, cv2, numpy as np, pandas as pd
from ultralytics import YOLO
from pathlib import Path
from difflib import SequenceMatcher
import warnings

# Mengabaikan peringatan umum
warnings.filterwarnings("ignore")

# --------------------------------------------------------------
# 📂 Direktori hasil (Menggunakan 'SAVE_BASE' dari CODE 1)
# --------------------------------------------------------------
if 'SAVE_BASE' not in globals():
    print("⚠️ Variabel 'SAVE_BASE' tidak ditemukan. Harap jalankan 'CODE 1' terlebih dahulu.")
    SAVE_BASE = "/content/drive/MyDrive/DISERTASI/RISET_PLANKTON/SEMINAR_KELAYAKAN_MULTI/UJIAN_KELAYAKAN/FALLBACK_DIR"

IOU_V5_DIR = Path(SAVE_BASE) / "IOU_AUTO_MATCH_V5_FINAL_RT-DETR"
IOU_V5_DIR.mkdir(parents=True, exist_ok=True)
print(f"📂 Hasil akan disimpan di: {IOU_V5_DIR}")

# --------------------------------------------------------------
# 🔹 Fungsi bantu
# --------------------------------------------------------------
def yolo_to_xyxy(b, w, h):
    x1 = (b[0] - b[2]/2) * w
    y1 = (b[1] - b[3]/2) * h
    x2 = (b[0] + b[2]/2) * w
    y2 = (b[1] + b[3]/2) * h
    return [x1, y1, x2, y2]

def box_iou(a, b):
    xA, yA = max(a[0], b[0]), max(a[1], b[1])
    xB, yB = min(a[2], b[2]), min(a[3], b[3])
    inter = max(0, xB-xA) * max(0, yB-yA)
    areaA = (a[2]-a[0])*(a[3]-a[1])
    areaB = (b[2]-b[0])*(b[3]-b[1])
    union = areaA + areaB - inter
    return inter/union if union > 0 else 0

# --------------------------------------------------------------
# 🔹 Evaluasi IoU (Versi Perbaikan v6)
# --------------------------------------------------------------
def evaluate_iou_v6(model_name, model_path, data_yaml): # Ganti nama fungsi
    print(f"\n{'='*100}\n🚀 Evaluasi IoU Auto-Match (V6-Stabil): {model_name}\n{'='*100}")

    try:
        model = YOLO(model_path)
    except Exception as e:
        print(f"❌ Gagal memuat model {model_name} dari {model_path}: {e}")
        return None, None

    # 1. Dapatkan kelas Validator yang benar
    try:
        ValidatorClass = model.task_map[model.task]['validator']
        print(f"ℹ️ {model_name}: Menggunakan Validator untuk tugas: {model.task}")
    except (AttributeError, KeyError):
        print(f"❌ Gagal mengambil ValidatorClass dari 'model.task_map'.")
        return None, None

    # 2. Siapkan argumen untuk validator (menggunakan akses dictionary)
    val_args = model.args.copy()
    val_args['data'] = data_yaml
    val_args['split'] = "test"
    val_args['imgsz'] = 640 # Fork lama ini seharusnya menerima integer
    val_args['save'] = True
    val_args['save_txt'] = True
    val_args['save_conf'] = True
    val_args['verbose'] = True
    val_args['project'] = "/content/ultralytics/runs/detect"
    val_args['name'] = f"val_{model_name.replace(' ', '_')}"

    # CATATAN: Jika ini crash dengan 'subscriptable', ubah baris di atas menjadi:
    # val_args['imgsz'] = [640, 640]

    # 3. Buat instance validator
    try:
        validator = ValidatorClass(args=val_args)
    except Exception as e:
         print(f"❌ Gagal membuat instance Validator untuk {model_name}: {e}")
         return None, None

    # 4. Jalankan validator.
    try:
        validator()
    except Exception as e:
        print(f"❌ Gagal saat menjalankan validator() untuk {model_name}: {e}")
        print("   Pastikan path di data.yaml (mis: 'test:') sudah benar.")
        return None, None

    # 5. Akses direktori penyimpanan prediksi
    pred_dir = validator.save_dir / "labels"

    if not pred_dir.exists():
        print(f"⚠️ {model_name}: folder prediksi tidak ditemukan di {pred_dir}")
        return None, None

    # 6. Ambil path 'test/images' dari data yang dimuat validator
    try:
        test_img_path = Path(validator.data['test'])
    except KeyError:
        print(f"❌ Gagal! 'test' key tidak ditemukan di 'validator.data'.")
        return None, None

    # 7. Turunkan path 'test/labels' dari path 'test/images'
    gt_dir = test_img_path.parent / 'labels'

    if not gt_dir.exists():
        print(f"❌ Gagal! Direktori Ground Truth tidak ditemukan di: {gt_dir}")
        return None, None

    # 8. Ambil daftar file GT secara manual
    gt_files_list = list(gt_dir.glob("*.txt"))

    if not gt_files_list:
        print(f"⚠️ {model_name}: Tidak ada file label (.txt) ditemukan di {gt_dir}")
        return None, None

    print(f"ℹ️ {model_name}: Menemukan {len(gt_files_list)} file ground truth di {gt_dir}")
    print(f"ℹ️ {model_name}: File prediksi disimpan di {pred_dir}")

    iou_per_class, matched, skipped = {}, 0, 0

    for gt_file in gt_files_list:
        gt_name = gt_file.stem
        match = max(list(pred_dir.glob("*.txt")), key=lambda p: SequenceMatcher(None, p.stem, gt_name).ratio(), default=None)

        if not match or SequenceMatcher(None, match.stem, gt_name).ratio() < 0.3:
            skipped += 1
            continue

        try:
            gt = np.loadtxt(gt_file, ndmin=2)
            pr = np.loadtxt(match, ndmin=2)
        except Exception as e:
            skipped += 1
            continue

        if gt.size == 0 or pr.size == 0:
            skipped += 1
            continue

        img_base = gt_file.as_posix().replace(os.sep + "labels" + os.sep, os.sep + "images" + os.sep).rsplit(".",1)[0]
        img_path = next((img_base + ext for ext in [".jpg", ".png", ".jpeg", ".JPG", ".PNG", ".JPEG"]
                         if os.path.exists(img_base + ext)), None)

        if not img_path:
            skipped += 1
            continue

        img = cv2.imread(img_path)
        if img is None:
            skipped += 1
            continue
        h, w = img.shape[:2]

        for p in pr:
            cls_p = int(round(p[0]))
            box_p = yolo_to_xyxy(p[1:5], w, h)
            gts = gt[np.isclose(gt[:, 0], cls_p, atol=1e-3)]

            if len(gts) == 0:
                continue

            ious = [box_iou(box_p, yolo_to_xyxy(g[1:5], w, h)) for g in gts]
            best_iou = max(ious)
            iou_per_class.setdefault(cls_p, []).append(best_iou)

        matched += 1

    print(f"✅ {model_name}: {matched} file cocok, {skipped} dilewati.")

    if not iou_per_class:
        print(f"⚠️ {model_name}: Tidak ada IoU valid ditemukan.")
        return None, None

    # Buat DataFrame
    df = pd.DataFrame({
        "Class_ID": list(iou_per_class.keys()),
        "Class_Name": [CLASS_NAMES[c] if c < len(CLASS_NAMES) else f"class_{c}" for c in iou_per_class.keys()],
        "IoU_mean": [np.mean(v) for v in iou_per_class.values()],
        "IoU_std": [np.std(v) for v in iou_per_class.values()],
        "IoU_50+": [np.mean(np.array(v) >= 0.5) for v in iou_per_class.values()],
        "IoU_75+": [np.mean(np.array(v) >= 0.75) for v in iou_per_class.values()],
        "Model": model_name
    })

    out_csv = IOU_V5_DIR / f"iou_per_class_{model_name}.csv"
    df.to_csv(out_csv, index=False, float_format="%.4f")
    print(f"✅ Disimpan: {out_csv}")

    mean_iou = df["IoU_mean"].mean()
    df_global = pd.DataFrame({
        "Model": [model_name],
        "IoU_mean_global": [mean_iou],
        "IoU_50_global": [df["IoU_50+"].mean()],
        "IoU_75_global": [df["IoU_75+"].mean()]
    })
    df_global.to_csv(IOU_V5_DIR / f"iou_global_{model_name}.csv", index=False, float_format="%.4f")
    return df, mean_iou

# --------------------------------------------------------------
# 🚀 Jalankan untuk semua model
# --------------------------------------------------------------
summary = []
print("\n" + "="*100)
print("🚀 MEMULAI EKSEKUSI SEMUA MODEL (V6-Stabil)")
print("="*100)

if 'model_paths' not in globals() or 'DATA_YAML' not in globals():
    print("❌ Gagal! Variabel 'model_paths' atau 'DATA_YAML' tidak ditemukan.")
else:
    for model_name, model_path in model_paths.items():
        if not Path(model_path).exists():
            print(f"\n⚠️ Melewatkan {model_name}: File model tidak ditemukan di {model_path}")
            continue

        df_iou, mean_iou = evaluate_iou_v6(model_name, model_path, DATA_YAML) # Ganti nama fungsi

        if mean_iou is not None:
            summary.append({"Model": model_name, "IoU_mean": mean_iou})

if summary:
    df_summary = pd.DataFrame(summary).sort_values("IoU_mean", ascending=False)
    df_summary.to_csv(IOU_V5_DIR / "iou_summary_all_models.csv", index=False, float_format="%.4f")

    print("\n" + "="*100)
    print("📊 Rekap Global IoU (Berhasil)")
    print("="*100)
    try:
        from google.colab.data_table import display
        display(df_summary)
    except ImportError:
        print(df_summary.to_string(index=False))
else:
    print("\n" + "="*100)
    print("❌ Tidak ada model yang berhasil dievaluasi.")
    print("="*100)

print(f"\n📂 Semua hasil disimpan di: {IOU_V5_DIR}")

**Comment by glenn-jocher** [comment]
Thanks for the details—no need to use `ValidatorClass` or filename heuristics. Save predictions with `val` and compute IoU directly against GT using our metrics helpers; this works the same for YOLOv8 and RT-DETR since the label format is identical.

```python
from ultralytics import YOLO
from ultralytics.utils.metrics import bbox_iou  # xywh or xyxy IoU
from pathlib import Path
import torch, numpy as np, yaml

data_yaml = "path/to/data.yaml"
name = "iou_eval"

model = YOLO("path/to/model.pt")  # yolov8*.pt or rtdetr*.pt
model.val(data=data_yaml, split="test", imgsz=640,
          save=True, save_txt=True, save_conf=True,
          project="runs/detect", name=name, exist_ok=True)

pred_dir = Path("runs/detect") / name / "labels"
with open(data_yaml) as f:
    d = yaml.safe_load(f)
gt_dir = Path(d.get("test", d.get("val"))).with_name("labels")

ious = []
for gt_path in gt_dir.glob("*.txt"):
    pr_path = pred_dir / gt_path.name
    if not pr_path.exists():
        continue
    gt = np.loadtxt(gt_path, ndmin=2)[:, :5]         # [cls, x, y, w, h] (normalized)
    pr = np.loadtxt(pr_path, ndmin=2)[:, :5]         # [cls, x, y, w, h] (conf dropped)
    for c in np.unique(pr[:, 0]).astype(int):
        a = torch.tensor(pr[pr[:, 0] == c, 1:5])     # preds (N,4) xywh
        b = torch.tensor(gt[gt[:, 0] == c, 1:5])     # gts   (M,4) xywh
        if len(a) and len(b):
            i = bbox_iou(a[:, None, :], b[None, :, :], xywh=True).max(dim=1).values
            ious.extend(i.tolist())

miou = float(torch.tensor(ious).mean()) if ious else float("nan")
print(f"mean IoU across predictions: {miou:.4f}")
```

Docs: see the Model Validation guide for saving predictions in `val` ([Model Validation with Ultralytics YOLO](https://docs.ultralytics.com/modes/val/)) and the IoU utilities you can call directly ([bbox_iou and related metrics APIs](https://docs.ultralytics.com/reference/utils/metrics/#bbox_iou)). For matching internals, the validator’s matcher is described here ([match_predictions reference](https://docs.ultralytics.com/reference/engine/validator/#ultralytics.engine.validator.BaseValidator.match_predictions)).
